xml/article_nfs_storage.xml

```diff
@@ -278,6 +268,41 @@
        </callout>
       </calloutlist>
      </step>
+     <step>
+      <para>
+       Create the file <filename>/etc/drbd.d/nfs-state.res</filename> with the
+       following contents. The differences from <filename>/etc/drbd.d/nfs-share.res</filename>
+       are shown in bold:
+      </para>
+<screen>resource nfs_<emphasis role="bold">state</emphasis> {
+   device /dev/<emphasis role="bold">drbd2 minor 2</emphasis>;
+   disk   /dev/nfs/<emphasis role="bold">state</emphasis>;
+   meta-disk internal;
+
+   net {
+      protocol  C;
+      fencing resource-and-stonith;
+   }
+
+   handlers {
+      fence-peer "/usr/lib/drbd/crm-fence-peer.9.sh";
+      after-resync-target "/usr/lib/drbd/crm-unfence-peer.9.sh";
```

Comment:
Hold on, 

The resource-and-stonith with crm-fence-peer.sh doesn't work well for multiple DRBD resources. With that, I incline to combine both resources into one, while still leave two volumes one for nfs-share, one for nfs-state.

That says, the design for multiple DRBD resources will need have more advanced consideration and trade-off, and might not fit this preliminary document.

Bottom line, the whole setup need more rigorous testing.

Reply:
I may be misunderstanding this comment.  We often have people using more than 2 of more file systems, shared via NFS, in one cluster.  If certain things don't work well for multiple DRDB resources, then that could be a problem as well, even without considering the "state" resource.  I'd like our steps to be scalable.  With multiple files systems shared via NFS, is there a way to combine those all into one resource as well, for purposes of this concern?

Reply:
Can we just leave out `resource-and-stonith` and `crm-fence-peer.sh`? What's the impact of not having them?

They're not included in the example `.res` files you attached in the bug, but I kept them from the original doc because they looked important :shrug: but that may have been a misjudgement on my part. 

I think I need to completely revisit the `.res` configs. What I did before was keep the configs from the original doc and change what seemed necessary based on your example configs. But now I realise I should have started by completely replacing the old config with your new example config, and then checked if anything from the old config should be added back.

Reply:
Now, this discussion start pulling in multiple advanced DRBD factors to be considered alone, 

> I'd like our steps to be scalable. With multiple files systems shared via NFS, is there a way to combine those all into one resource as well, for purposes of this concern?

1. Scalability is a good angle. Yes, DRBD is flexible to make it work to add more DRBD volumes with individual filesystem on each. However, a number of steps is needed with caution. Besides that, because of such flexibility, it can have variants for different purposes. For example, if "scalable" its self means the minimal interruption to the service to scale online, that will add further complexity to administration .

> Can we just leave out resource-and-stonith and crm-fence-peer.sh? What's the impact of not having them?

Good question. This is the more in-depth of DRBD internals. `man drbd.conf` tells more. Put it simple, resource-and-stonith is the fast approach to react on potential split-brain in the DRBD cluster. But it comes with limitation just like we are facing. 

2. Simplicity is another angle I think this DOC trying to be. With that in mind, yes, have "multiple DRBD resources without resource-and-stonith" need much less steps to scale to multiple filesystems. 

3. Resilience level can be different for DRBD Clusters . Besides the "resource-and-stonith" approach, also there are other approaches,   eg. "automatic split brain policy" in my example attached in the bug. Even, some customers equip both.

Circle back to the motivation of this effort to improve NFS DRBD use case to be more resilient on failures. I slightly incline to setup "resource-and-stonith".  

In the next few days, I could do more homework and come up with the appropriate minimum steps about how to add DRBD volumes for this NFS use case. It could show us its complexity level, and judge if worth to DOC it. Or, lucky enough to find a better way out.

Anyway, it is still open for more comments, more discussion, and debate.

Reply:
Here below is my example steps to scale one more filesystem into the running cluster. That sounds doable for this DOC.

"""

    1. crm cluster run "lvcreate -n share2 -L 1G nfs"

    2. Update nfs.res to add new volume. Caution on: vol-number, minor-X, /dev/nfs/shareX
      >     volume 2 {
      >         device           /dev/drbd3;
      >         disk             /dev/nfs/share2;

      cysnc2 the configuration.

    3. crm cluster run "drbdadm create-md nfs/2 --force"

    4. crm cluster run "drbdadm adjust nfs"

    5. drbdadm new-current-uuid --clear-bitmap nfs/2

    6. mkfs.ext4 /dev/drbd3

    7. crm config edit to add new filesystem and exportfs etc.

"""

Here are the reference CIB and drbd .res files: 
[nfs-1more-share.0322.cib.txt](https://github.com/SUSE/doc-sleha/files/11041218/nfs-1more-share.0322.cib.txt)
[nfs-1more-share.0322.res.txt](https://github.com/SUSE/doc-sleha/files/11041220/nfs-1more-share.0322.res.txt)

NOTE:  

a) drbd nfs.res consolidate all into multi-volume drbd devices. Hence <resources/vol-number> designation is used often now. 

b)  For the cluster configuration CIB, I stretch my NFS knowledge again, and fail the attempt to isolate the failure of one filesystem to leave other nfs shares keep providing service. Basically, I restore CIB to the plain group resource type. That says, stopping one filesystem will impact the whole cluster, for example.

Reply:
Thanks for those steps, Roger. I think that's the ideal way to present it: as a final section showing an example of how to add more filesystems to the config. 

Regarding groups:

From what I understand, resources in a group are started in the order they're added to the group (and stopped in reverse order). This is convenient for addressing point 3 in the TID, where the resources need to be started in a specific order. The group setup is also much simpler to present in the doc, so I'm on board with using a group.

~However, when we create the new filesystem and exportfs, does that mean that adding them to the group last means they'll be started last? How do we adjust the group so that the new filesystem and exportfs slot in to the appropriate starting order?~ 

Actually, `crm help configure modgroup` had the answer. We can use `before` or `after` to specify where the new resources go in the group. Solved :+1: